biz.aQute.repository/src/aQute/bnd/repository/maven/provider/IndexFile.java

```diff
@@ -13,403 +11,527 @@
-		public long		lastModified;
-		public Archive	archive;
-		public boolean	merged;
-		String			error;
-		Promise<File>	promise;
-		Resource		resource;
-
-		public synchronized Resource getResource() {
-			if (resource == null) {
-
-				try {
-					File f = promise.getValue();
-					ResourceBuilder rb = new ResourceBuilder();
-					rb.addFile(f, f.toURI());
-					resource = rb.build();
-				} catch (Exception e) {
-					reporter.exception(e, "Failed to get file for %s", archive);
-					resource = ResourceUtils.DUMMY_RESOURCE;
-				}
-			}
-			return resource;
-		}
-	}
-
-	private final ConcurrentMap<Archive, Promise<File>>	promises	= new ConcurrentHashMap<>();
-	final ConcurrentMap<Archive, BundleDescriptor>		descriptors	= new ConcurrentHashMap<>();
-	final File											indexFile;
-	final File											cacheDir;
-	private final IMavenRepo							repo;
-	private final Reporter								reporter;
-	private long										lastModified;
-	private AtomicBoolean								refresh		= new AtomicBoolean();
-	private final ReadWriteLock							lock		= new ReentrantReadWriteLock();
-	private final PromiseFactory						promiseFactory;
-
-	IndexFile(Reporter reporter, File file, IMavenRepo repo, PromiseFactory promiseFactory) throws Exception {
+	private final static Logger			logger		= LoggerFactory.getLogger(IndexFile.class);
+
+	final File							indexFile;
+	final IMavenRepo					repo;
+	final Reporter						reporter;
+	final PromiseFactory				promiseFactory;
+	final Map<Archive, Resource>		archives	= new ConcurrentHashMap<>();
+	final String[]						multi;
+
+	private long						lastModified;
+	private long						last		= 0L;
+	volatile Promise<BridgeRepository>	bridge;
+
+	/*
+	 * Constructor
+	 */
+	IndexFile(Reporter reporter, File file, IMavenRepo repo, PromiseFactory promiseFactory, String... multi)
+		throws Exception {
 		this.reporter = reporter;
 		this.indexFile = file;
 		this.repo = repo;
-		this.cacheDir = new File(indexFile.getParentFile(), indexFile.getName() + ".info");
 		this.promiseFactory = promiseFactory;
+		this.multi = multi;
+		bridge = null;
 	}
 
+	/*
+	 * Open, will refresh this repository, which will load the file
+	 */
 	void open() throws Exception {
-		loadIndexFile();
-		sync();
+		this.bridge = load();
 	}
 
-	private void sync() throws Exception {
-		List<Promise<File>> sync = new ArrayList<>(promises.size());
-		for (Iterator<Entry<Archive, Promise<File>>> i = promises.entrySet()
-			.iterator(); i.hasNext();) {
-			Entry<Archive, Promise<File>> entry = i.next();
-			i.remove();
-			sync.add(entry.getValue()
-				.onFailure(failure -> reporter.exception(failure, "Failed to sync %s", entry.getKey())));
-		}
-		promiseFactory.all(sync)
-			.getFailure(); // block until all promises resolved
-	}
+	/*
+	 * Persistently add an archive to the index file and refresh. If the archive
+	 * is not kept then this is a noop
+	 */
+	synchronized void add(Archive archive) throws Exception {
```

Comment:
I am not sure why some methods are synchronized and other which seem to manipulate the same archives map are not. It also seems like there is an awful lot done inside the synchronized block.

Reply:
add and the removes methods must be synchronized. Hard work is done in update() that needs to know the delta. If the add/remove methods could be called overlapping the delta could be off. Quite theoretical though. The update does the parsing synchronous if it is local as an optimization. Could do it in the background but that seems more overhead than gain. Any download code is done in another thread.